dotnet/src/SemanticKernel/Reliability/DefaultHttpRetryHandler.cs

```diff
@@ -0,0 +1,259 @@
+                {
+                    return response;
+                }
+
+                // Drain response content to free connections. Need to perform this
+                // before retry attempt and before the TooManyRetries ServiceException.
+                if (response.Content != null)
+                {
+#if NET5_0_OR_GREATER
+                    await response.Content.ReadAsByteArrayAsync(cancellationToken).ConfigureAwait(false);
+#else
+                    await response.Content.ReadAsByteArrayAsync().ConfigureAwait(false);
+#endif
+                }
+
+                reason = response.StatusCode.ToString();
+
+                if (retryCount >= this._config.MaxRetryCount)
+                {
+                    this._log.LogError(
+                        "Error executing request, max retry count reached. Reason: {0}", reason);
+                    return response;
+                }
+
+                // If the retry delay is longer than the total timeout, then we'll
+                // just return
+                if (!this.HasTimeForRetry(start, retryCount, response, out waitFor))
+                {
+                    this._log.LogError(
+                        "Error executing request, max total retry time reached. Reason: {0}", reason);
+                    return response;
+                }
+            }
+            catch (Exception e) when ((this.ShouldRetry(e) || this.ShouldRetry(e.InnerException)) &&
+                                      retryCount < this._config.MaxRetryCount &&
+                                      this.HasTimeForRetry(start, retryCount, response, out waitFor))
+            {
+                reason = e.GetType().ToString();
+            }
+
+            // If the request requires a retry then we'll retry
+            this._log.LogWarning(
+                "Error executing action [attempt {0} of {1}]. Reason: {2}. Will retry after {3}ms",
+                retryCount + 1,
+                this._config.MaxRetryCount,
+                reason,
+                waitFor.TotalMilliseconds);
+
+            // Clone request with CloneAsync before retrying
+            // Do not dispose this request as that breaks the request cloning
+#pragma warning disable CA2000
+            request = await CloneAsync(request);
+#pragma warning restore CA2000
+
+            // Increase retryCount
+            retryCount++;
+
+            // Delay
+            await this._delayProvider.DelayAsync(waitFor, cancellationToken).ConfigureAwait(false);
+        }
+    }
+
+    #region private
+
+    private readonly HttpRetryConfig _config;
+    private readonly ILogger _log;
+    private readonly IDelayProvider _delayProvider;
+    private readonly ITimeProvider _timeProvider;
+
+    private TimeSpan GetWaitTime(int retryCount, HttpResponseMessage? response)
+    {
+        var retryAfter = response?.Headers.RetryAfter?.Date.HasValue == true
+            ? response?.Headers.RetryAfter?.Date - DateTimeOffset.Now
```

Comment:
how is locale handled? should this use UtcNow? why not use ITimeProvider instead of DateTimeOffset.Now?

Reply:
> how is locale handled?
The RetryConditionHeaderValue and RetryAfter properties in the System.Net.Http.Headers namespace are not affected by locale settings. These properties are used to represent the value of the Retry-After header in an HTTP response.

The Retry-After header is used to inform the client how long it should wait before retrying a failed request. The value of the Retry-After header can be either an HTTP-date or an integer representing the number of seconds to wait before retrying.

Since the value of the Retry-After header is either an HTTP-date or an integer, it is not affected by locale settings. An HTTP-date is always formatted according to the rules specified in RFC 1123, which uses the English language and the GMT time zone. Similarly, an integer representing the number of seconds to wait is language-agnostic.

Therefore, the RetryConditionHeaderValue and RetryAfter properties do not have any special handling for locale.


> why not use ITimeProvider instead of DateTimeOffset.Now

Good call.